### DISCUSSION TITLE
Dev Webserver works, gunicorn webserver throws "Static Media Failure"

### DISCUSSION THREAD

**Discussion by halb9** [netbox-community/netbox#8598] [Q&A]
Hi there, 

I'm having trouble getting netbox up and running. 

Using Ubuntu 20.04 LTS as the underlying OS I followed the steps in the documentation precisely. 

1. Install [PostgreSQL](https://netbox.readthedocs.io/en/stable/installation/1-postgresql/)
(Postgres is running on a different host. Not that I think that has anything to do with my issue. But wanted to mention it regardless.)
2. Install [Redis](https://netbox.readthedocs.io/en/stable/installation/2-redis/)
3. Install [NetBox](https://netbox.readthedocs.io/en/stable/installation/3-netbox/)
(I'm cloning the repo via git at this point). 
It all looks good. At the end of this step I'm asked to test the application via 
`python3 manage.py runserver 0.0.0.0:8000 --insecure`
Opening the IP of the host netbox is running on with my local browser works just fine. I can login with the user I created earlier. All seems good. 
I continue with 
4. [Gunicorn](https://netbox.readthedocs.io/en/stable/installation/4-gunicorn/)
Again, installation went fine. No errors. No hickups. gunicorn is running as a service on TCP 8001. 
However when I now try to open it via my browser I get the following message. 
![static-media-failure](https://user-images.githubusercontent.com/18451020/153271262-715c839a-838f-41c7-8df8-845ab3290c87.png)

```
Static Media Failure
The following static media file failed to load: netbox-dark.css
Check the following:

manage.py collectstatic was run during the most recent upgrade. This installs the most recent iteration of each static file into the static root path.
The HTTP service (e.g. nginx or Apache) is configured to serve files from the STATIC_ROOT path. Refer to the installation documentation for further guidance.
STATIC_URL: /static/
The file netbox-dark.css exists in the static root directory and is readable by the HTTP process.
Click here to attempt loading NetBox again.
```

I take a look at the server via CLI and files are all there. 

I doublechecked my setup process and I'm very confident I did not miss any steps. However I still might be wrong. 

I tried the search of this repo of course. But there were either [no answers](https://github.com/netbox-community/netbox/discussions/7713) or the issue was [similar but not identical](https://github.com/netbox-community/netbox/discussions/7511). 
There is [one issue with a marked solution](https://github.com/netbox-community/netbox/discussions/7831), yet in there nginx is used. 
What are the required steps at this point so that gunicorn in itself works?
Any help is appreciated. 

Have a great day! Thanks!

**Comment by candlerb** [comment]
You should not be talking to gunicorn directly on port 8001 (indeed it should be bound to 127.0.0.1 so it doesn't permit remote access)

You need a proxy in front of it (Apache or Nginx), and you talk to that, on port 80 or 443 usually.  The proxy serves static files, and anything that's not under `/static/` it proxies to gunicorn on port 8001.

**Comment by halb9** [comment]
Hi candlerb, 

thanks for taking the time. I appreciate that!

Maybe I wasn't clear enough what my intentions are. It is not my goal to talk to gunicorn directly in production. I just wanted to verify this installation step at this point. 
In the end I'd like to have nginx as a reverse proxy not on the same machine but on a different host 'proxieng' traffic to netbox/gunicorn. 

You said gunicorn should be bound to 127.0.0.1 with no remote access. Why? 
I'm coming from the network perspective. I would like to know what services are talking to each other. 

According to the documentation nginx talks to gunicorn, gunicorn talks to netbox etc. 
![image](https://user-images.githubusercontent.com/18451020/153365096-c110b85a-a359-4b3a-a133-74d61910157c.png)

When nginx as a reverse proxy is running on a different host, gunicorn can not be bound to 127.0.0.1. Right? 
In my scenario nginx talks to gunicorn. Which means gunicorn should accept outside connections. Which means I should be able to access it via my browser. For functionality testing. There is no point going forward when this step fails. 

I'm looking forward to your feedback. Maybe I'm doing something wrong and/or stupid :D

**Reply by candlerb** [reply]
(Aside: that diagram appears to be missing a piece, the `media` items. These can be stored on the local filesystem alongside Netbox, or in remote storage like S3)

> You said gunicorn should be bound to 127.0.0.1 with no remote access. Why?

That's the standard way to deploy Netbox.  You can do it differently, but you'll be making life harder for yourself.  In order to expose gunicorn directly, then you should first ensure that it has some sort of ACL that prevents direct access except from the front-end proxy's IP address.  Then you'll have to come up with your own solution for `/static/` assets because, as you've found and by design, gunicorn doesn't serve those.  For example, you could copy them to the remote host where Nginx is running, or share them with NFS.

If you need a remote proxy for some other purpose - e.g. restricting public access - then you'll probably find it easier to chain proxies, i.e. run a local one on the Netbox server as well.  There is a minor issue when doing that, IIRC due to Django not dealing with the X-Forwarded-Host header having multiple values, but I've had it working before and I can dig out the workaround if you need it.

There is one place where I have implemented something a bit like what you're looking at: I have Apache running on a monitoring server, and Netbox running inside an lxd container on that server.  To avoid the double-proxy I just fetched the static assets directly out of the container's filesystem:

```
        Alias /static /var/lib/lxd/containers/netbox/rootfs/opt/netbox/netbox/static/
        <Directory /var/lib/lxd/containers/netbox/rootfs/opt/netbox/netbox/static/>
                Options Indexes FollowSymLinks MultiViews
                AllowOverride None
                Require all granted
        </Directory>

        <Location /static>
                ProxyPass !
        </Location>

        ProxyPreserveHost On
        RequestHeader set X-Forwarded-Proto "https"
        ProxyPass / http://10.112.76.119:8001/
        ProxyPassReverse / http://10.112.76.119:8001/
```

That works nicely, and gives me the benefit of sharing the proxy between multiple applications all nicely boxed up in their lxd containers.  But if the proxy were on a different physical server, I'd need to share the static assets some other way.

**Reply by halb9** [reply]
Thanks for the detailed answer. As this is not my area of expertise I'll have to do a bit of research and testing first before I come back to you. 
For now thank you very much!

**Reply by candlerb** [reply]
I found the workaround for the multi-hop proxying.  On the outermost proxy, which was running Apache, I did this:

```
# Keep the Host: header
ProxyPreserveHost On

# Don't add X-Forwarded-Host: as this breaks when proxying multiple times
ProxyAddHeaders Off
RequestHeader setifempty X-Forwarded-For "expr=%{REMOTE_ADDR}"
RequestHeader setifempty X-Forwarded-Proto "expr=%{REQUEST_SCHEME}"
```

If you do something like this, then you can do a "standard" installation of Netbox, with its own local proxy as well.  The local proxy will be responsible for serving static assets, and this means that when you upgrade Netbox in future the static assets will remain in sync with no extra work.

**Reply by NavyWings** [reply]
> I found the workaround for the multi-hop proxying. On the outermost proxy, which was running Apache, I did this:
> 
> ```
> # Keep the Host: header
> ProxyPreserveHost On
> 
> # Don't add X-Forwarded-Host: as this breaks when proxying multiple times
> ProxyAddHeaders Off
> RequestHeader setifempty X-Forwarded-For "expr=%{REMOTE_ADDR}"
> RequestHeader setifempty X-Forwarded-Proto "expr=%{REQUEST_SCHEME}"
> ```
> 
> If you do something like this, then you can do a "standard" installation of Netbox, with its own local proxy as well. The local proxy will be responsible for serving static assets, and this means that when you upgrade Netbox in future the static assets will remain in sync with no extra work.

Thanks candleb. I'm not proficient in this area either. So do you mean specifying that in apache2.conf while keeping netbox.conf a standard install? If so, would it go in:

<Directory />
        Options FollowSymLinks
        AllowOverride None
        Require all denied
</Directory>

<Directory /usr/share>
        AllowOverride None
        Require all granted
</Directory>

<Directory /var/www/>
        Options Indexes FollowSymLinks
        AllowOverride None
        Require all granted
</Directory>

**Reply by candlerb** [reply]
The outer proxy is your system, and you'll need to configure it in whatever way makes sense for you.  I have no idea what virtualhosts or other configuration you have, so I can't point you to an exact line.  Ask a local system administrator if you need help.

**Reply by NavyWings** [reply]
> The outer proxy is your system, and you'll need to configure it in whatever way makes sense for you. I have no idea what virtualhosts or other configuration you have, so I can't point you to an exact line. Ask a local system administrator if you need help.

Thanks! Just got it up and running. Somehow I think it broke my SSL but that's another issue.

**Reply by Sh4wn9** [reply]
@halb9 What was your solution as I have followed the documentation and getting same Static Media Failure. Thanks in advance!

**Reply by alihyder-s** [reply]
I am also getting the same error of static media failure. Any solution? Followed all the steps as per guide for the installation on Ubuntu Server20.04.
